From 99f8ab16f63f11953c06aacb33809c967dc5a4de Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Fri, 12 Sep 2014 01:56:50 -0400
Subject: [drm] i915: mst topology dumper in debugfs (v0.2)

Message-id: <1410487019-24661-13-git-send-email-airlied@redhat.com>
Patchwork-id: 93929
O-Subject: [RHEL7.1 BZ1140440 PATCH 12/21] i915: mst topology dumper in debugfs (v0.2)
Bugzilla: 1140440
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Rob Clark <rclark@redhat.com>

bugzilla: 1140440
upstream: 11bed958b72e15fd12d78c30ce49047b94817840

use the mst helper code to dump the topology in debugfs.

v0.2: drop is_mst check - as we want to dump other info

Signed-off-by: Dave Airlie <airlied@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/i915_debugfs.c b/drivers/gpu/drm/i915/i915_debugfs.c
index b8c6892..348054d 100644
--- a/drivers/gpu/drm/i915/i915_debugfs.c
+++ b/drivers/gpu/drm/i915/i915_debugfs.c
@@ -2383,6 +2383,28 @@ struct pipe_crc_info {
  enum pipe pipe;
 };
 
+static int i915_dp_mst_info(struct seq_file *m, void *unused)
+{
+ struct drm_info_node *node = (struct drm_info_node *) m->private;
+ struct drm_device *dev = node->minor->dev;
+ struct drm_encoder *encoder;
+ struct intel_encoder *intel_encoder;
+ struct intel_digital_port *intel_dig_port;
+ drm_modeset_lock_all(dev);
+ list_for_each_entry(encoder, &dev->mode_config.encoder_list, head) {
+  intel_encoder = to_intel_encoder(encoder);
+  if (intel_encoder->type != INTEL_OUTPUT_DISPLAYPORT)
+   continue;
+  intel_dig_port = enc_to_dig_port(encoder);
+  if (!intel_dig_port->dp.can_mst)
+   continue;
+
+  drm_dp_mst_dump_topology(m, &intel_dig_port->dp.mst_mgr);
+ }
+ drm_modeset_unlock_all(dev);
+ return 0;
+}
+
 static int i915_pipe_crc_open(struct inode *inode, struct file *filep)
 {
  struct pipe_crc_info *info = inode->i_private;
@@ -3823,6 +3845,7 @@ static const struct drm_info_list i915_debugfs_list[] = {
  {"i915_pc8_status", i915_pc8_status, 0},
  {"i915_power_domain_info", i915_power_domain_info, 0},
  {"i915_display_info", i915_display_info, 0},
+ {"i915_dp_mst_info", i915_dp_mst_info, 0},
 };
 #define I915_DEBUGFS_ENTRIES ARRAY_SIZE(i915_debugfs_list)
 
-- 
1.7.1